db/src/models/game.rs

```diff
@@ -1081,36 +1068,142 @@ impl Game {
+                .map_or(vec![], |v| v.clone());
+            query = query.filter(
+                username
+                    .eq_any(all_both)
+                    .or(is_white.and(username.eq_any(all_white)))
+                    .or(is_black.and(username.eq_any(all_black)))
+                    .or(BoolExpressionMethods::or(
+                        is_white.and(white_won.clone()),
+                        is_black.and(black_won.clone()),
+                    )
+                    .and(username.eq_any(win_both)))
+                    .or(is_white
+                        .and(white_won.clone())
+                        .and(username.eq_any(win_white)))
+                    .or(is_black
+                        .and(black_won.clone())
+                        .and(username.eq_any(win_black)))
+                    .or(BoolExpressionMethods::or(
+                        is_white.and(black_won.clone()),
+                        is_black.and(white_won.clone()),
+                    )
+                    .and(username.eq_any(loss_both)))
+                    .or(is_white
+                        .and(black_won.clone())
+                        .and(username.eq_any(loss_white)))
+                    .or(is_black
+                        .and(white_won.clone())
+                        .and(username.eq_any(loss_black)))
+                    .or(is_draw.clone().and(username.eq_any(draw_both)))
+                    .or(is_draw
+                        .clone()
+                        .and(is_white.and(username.eq_any(draw_white))))
+                    .or(is_draw.and(is_black.and(username.eq_any(draw_black)))),
+            );
+        }
+        //Selected speeds
+        let speeds: Vec<_> = options.speeds.iter().map(|s| s.to_string()).collect();
+        query = query.filter(games::speed.eq_any(speeds));
+
+        query = query.order_by((games::updated_at.desc(), games::id.desc()));
 
-        if let (Some(last_updated_at), Some(last_id)) = (last_updated_at, last_game_id) {
+        //batching
+        if let Some(batch) = &options.current_batch {
             query = query.filter(diesel::BoolExpressionMethods::or(
-                games::updated_at.lt(last_updated_at),
+                games::updated_at.lt(batch.timestamp),
                 diesel::BoolExpressionMethods::and(
-                    games::updated_at.eq(last_updated_at),
-                    games::id.ne(last_id),
+                    games::updated_at.eq(batch.timestamp),
+                    games::id.ne(batch.id),
                 ),
-            ))
-        };
-
-        Ok(query
-            .limit(amount)
-            .select(games::all_columns)
-            .get_results(conn)
-            .await?)
+            ));
+        }
+        if let Some(batch_size) = options.batch_size {
+            query = query.limit(batch_size as i64);
+        }
+        Ok(query.select(games::all_columns).get_results(conn).await?)
     }
 
```

Comment:
Would be good if we could simplify / break this up into smaller functions although it might be complicated to do due to diesel types...Tbh I'm not sure I'd be able to do it myself.

If not simplify at least mark it with a  // TODO: ...